### PR TITLE
Refactor opening ride

### PR DESCRIPTION
**Pull request by spacek531**
This PR lays the groundwork for #18714 by reducing copied code and separating ::Simulate from ::Test.

### CODE REVIEW

File: src/openrct2/ride/Ride.cpp [hunk 1]
```diff
@@ -3889,21 +3889,8 @@ TrackElement* Ride::GetOriginElement(StationIndex stationIndex) const
     return nullptr;
 }
 
-ResultWithMessage Ride::Test(RideStatus newStatus, bool isApplying)
+ResultWithMessage Ride::ChangeStatusGetStartElement(CoordsXYE* trackElement)
```
Comment (Gymnasiast) [inline src/openrct2/ride/Ride.cpp:3892]: Use a reference here, not a pointer.

**Comment by spacek531** [comment]
I am marking as draft because I introduced a bug. I will fix the bug before marking as ready for review.

**Comment by spacek531** [comment]
The bug was that when building a station end and changing the status to simulate, the station indices would be corrupted and the ghost train would appear on the ghost station element. I do not know what the cause was but taking everything out of ::Simulate and putting it all back in again fixed it.

**Comment by Gymnasiast** [comment]
I have backported parts of this to https://github.com/OpenRCT2/OpenRCT2/pull/18867 , to make sure that they can be reviewed properly - it would have been much easier if the function you spun off had commits of their own.

File: src/openrct2/ride/Ride.cpp [hunk 2]
```diff
@@ -3881,77 +3881,83 @@ TrackElement* Ride::GetOriginElement(StationIndex stationIndex) const
...
     {
-        if (FindTrackGap(trackElement, &problematicTrackElement) && (newStatus != RideStatus::Simulating || IsBlockSectioned()))
-        {
-            ride_scroll_to_track_error(problematicTrackElement);
-            return { false, STR_TRACK_IS_NOT_A_COMPLETE_CIRCUIT };
-        }
+        return message;
+    }
+
+    if (FindTrackGap(trackElement, &problematicTrackElement) && IsBlockSectioned())
```
Comment (Gymnasiast) [inline src/openrct2/ride/Ride.cpp:3939]: Reverse the order - a block section check is cheaper than finding the track gap.

**Comment by Gymnasiast** [comment]
I have extensively rebased and modified the PR to improve it and reduce the diff. But I might split off more.

**Comment by Gymnasiast** [comment]
I’m finished refactoring the code. I went for a more cautious approach. The way the commits are split are also how I would like to see any PR going forward - individual commits that can stand on their own and make it easy to track what is going on.

**Comment by spacek531** [comment]
What did you do to split the PR like this? Soft reset and interactive commit?

**Comment by Gymnasiast** [comment]
Hard reset, and then I used copy-paste for the intermediate steps. But that was mostly because I wanted to rewrite bits of it. If you’re happy about the code, soft reset and interactive commit could be a good option (especially if you have an IDE that makes this easy, which I have).

**Review by Gymnasiast — APPROVED** [review]

**Pull request merged**